File: src/main/java/cookbook/backend/BudgetService.java
package cookbook.backend;

public class BudgetService {

    // A simple helper container to return multiple pieces of data to the frontend
    public static class BudgetResult {
        public boolean isSuccess;
        public String errorMessage;
        public String formattedBudget;

        public BudgetResult(boolean isSuccess, String errorMessage, String formattedBudget) {
            this.isSuccess = isSuccess;
            this.errorMessage = errorMessage;
            this.formattedBudget = formattedBudget;
        }
    }

    public static BudgetResult validateBudget(String budgetStr, double currentTotalCost) {
        if (budgetStr.isEmpty() || budgetStr.equals("Enter Budget")) {
            return new BudgetResult(false, "Please enter a budget!", null);
        }

        double numericNewBudget = 0.0;
        try {
            // Strip out "php", commas, and spaces to get the raw number
            numericNewBudget = Double.parseDouble(budgetStr.toLowerCase().replace("php", "").replace(",", "").trim());
        } catch (Exception ex) {
            return new BudgetResult(false, "Please enter a valid number!", null);
        }

        // Check against the current shopping cart cost
        if (currentTotalCost > 0.0 && numericNewBudget < currentTotalCost) {
            String errorMsg = "Too low! Current meal costs Php " + String.format("%.2f", currentTotalCost);
            return new BudgetResult(false, errorMsg, null);
        }

        // Format the valid budget correctly
        String finalBudget;
        if (!budgetStr.toLowerCase().startsWith("php")) {
            finalBudget = "Php " + String.format("%.2f", numericNewBudget);
        } else {
            finalBudget = budgetStr;
        }

        return new BudgetResult(true, null, finalBudget);
    }
}

File: app/src/main/java/com/cookbook/backend/BudgetService.kt
package com.cookbook.backend

object BudgetService {

    data class BudgetResult(
        val isValid: Boolean,
        val formattedBudget: String = "",
        val errorMessage: String = ""
    )

    fun validateBudget(input: String, currentTotalCost: Double = 0.0): BudgetResult {
        if (input.isBlank()) {
            return BudgetResult(false, errorMessage = "Please enter a budget.")
        }
        try {
            val cleaned = input.replace(Regex("[PhpPHPphp\\s,]"), "").trim()
            val amount = cleaned.toDouble()
            if (amount < 0) {
                return BudgetResult(false, errorMessage = "Budget cannot be negative.")
            }
            val formatted = "Php %.2f".format(amount)
            return BudgetResult(true, formattedBudget = formatted)
        } catch (_: NumberFormatException) {
            return BudgetResult(false, errorMessage = "Please enter a valid number.")
        }
    }
}
